devices/ble_hci/common-hal/_bleio/Characteristic.c

```
@@ -0,0 +1,228 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+
+#include "shared-bindings/_bleio/__init__.h"
+#include "shared-bindings/_bleio/Characteristic.h"
+#include "shared-bindings/_bleio/CharacteristicBuffer.h"
+#include "shared-bindings/_bleio/Descriptor.h"
+#include "shared-bindings/_bleio/PacketBuffer.h"
+#include "shared-bindings/_bleio/Service.h"
+
+#include "common-hal/_bleio/Adapter.h"
+#include "common-hal/_bleio/att.h"
+
+#define CCCD_NOTIFY   0x1
+#define CCCD_INDICATE 0x2
+
+
+void common_hal_bleio_characteristic_construct(bleio_characteristic_obj_t *self, bleio_service_obj_t *service, uint16_t handle, bleio_uuid_obj_t *uuid, bleio_characteristic_properties_t props, bleio_attribute_security_mode_t read_perm, bleio_attribute_security_mode_t write_perm, mp_int_t max_length, bool fixed_length, mp_buffer_info_t *initial_value_bufinfo) {
+    self->service = service;
+    self->uuid = uuid;
+    self->decl_handle = BLE_GATT_HANDLE_INVALID;
+    self->handle = BLE_GATT_HANDLE_INVALID;
+    self->props = props;
+    self->read_perm = read_perm;
+    self->write_perm = write_perm;
+    self->descriptor_list = mp_obj_new_list(0, NULL);
+    self->observer = mp_const_none;
+    self->user_desc = NULL;
+    self->cccd = NULL;
+    self->sccd = NULL;
+    self->value = mp_obj_new_bytes(initial_value_bufinfo->buf, initial_value_bufinfo->len);
+
+    const mp_int_t max_length_max = 512;
+    if (max_length < 0 || max_length > max_length_max) {
+        mp_raise_ValueError(translate("max_length must be <= 512"));
+    }
+    self->max_length = max_length;
+    self->fixed_length = fixed_length;
+
+    if (service->is_remote) {
```

Should we raise an exception when the Characteristic is remote? Do we raise NotImplementedErrors in the cases that aren't implemented yet?

---

Currently the scanning and connection routines raise `NotImplementedError`, so there's no way to even get to this path in the code, because there's no way to create a remote Characteristic. So I didn't bother to add raises in cases like this.